File: RemoveDuplicatesFromSortedList2.py
"""
Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

For example,
Given 1->2->3->3->4->4->5, return 1->2->5.
Given 1->1->1->2->3, return 2->3.
"""
from ListNode import ListNode


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def deleteDuplicates(self, head):
        if not head:
            return head
        dummy_head = ListNode(head.val - 1)
        dummy_head.next = head
        pre = dummy_head
        while pre:
            cur = pre.next
            if cur and cur.next and cur.val == cur.next.val:
                while cur and cur.next and cur.val == cur.next.val:
                    cur = cur.next
                pre.next = cur.next
                cur = pre.next
            else:
                pre = cur
        return dummy_head.next


if __name__ == '__main__':
    head = ListNode(1, 1, 1, 2, 3)
    print Solution().deleteDuplicates(head)

File: Java-By-Tag/src/main/RemoveDuplicatesFromSortedList2.java
package src.main;

/**
 * Created by yhf on 9/15/15.
 */

/**
 * Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.
 *
 * For example,
 * Given 1->2->3->3->4->4->5, return 1->2->5.
 * Given 1->1->1->2->3, return 2->3.
 */

public class RemoveDuplicatesFromSortedList2 {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0), cur = dummy;
        dummy.next = head;
        while (cur != null) {
            if (cur.next != null && cur.next.next != null && cur.next.val == cur.next.next.val) {
                ListNode temp = cur;
                cur = cur.next;
                while (cur.next != null && cur.val == cur.next.val) {
                    cur = cur.next;
                }
                temp.next = cur.next;
                cur = temp;
            } else {
                cur = cur.next;
            }
        }
        return dummy.next;
    }
}
